example/extensions/lib_subgraph/subgraph_lib.cc

```diff
@@ -0,0 +1,278 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ * \file subgraph_lib.cc
+ * \brief subgraph operator implementation library file
+ */
+
+#include <math.h>
+#include <iostream>
+#include <algorithm>
+#include "lib_api.h"
+
+MXReturnValue parseAttrs(std::map<std::string, std::string> attrs,
+                         int* num_in, int* num_out) {
+  *num_in = 1;
+  *num_out = 1;
+  if (attrs.count(SUBGRAPH_SYM_JSON)) {
+    // example of subgraph json parsing
+    JsonParser jp;
+    JsonVal val = jp.parse_to_json(attrs[SUBGRAPH_SYM_JSON]);
+    int input = 0;
+    for (auto &item : val.map[JsonVal("nodes")].list) {
+      if (item.map[JsonVal("op")].str == "null")
+        input++;
+    }
+    int output = val.map[JsonVal("heads")].list.size();
+    *num_in = input;
+    *num_out = output;
+  }
+  return MX_SUCCESS;
+}
+
+MXReturnValue inferType(std::map<std::string, std::string> attrs,
+                        std::vector<int> &intypes,
```

Comment:
Should  we set "intypes" to const?

Reply:
infer type/shape may set types for both inputs and outputs (it does a multi pass propagation forward and backward) so inputs could be set.